make Rails::Rack::Logger tag compute configurable

skipkayhil opened the pull request:
### Summary

For applications that use a structured/JSON logger, this is the first
change necessary to allow them to use the Tagged Logging feature in
Rails::Rack::Logger. Instead of returning an array of strings, a
configured tag computer could return a hash of fields to be merged into
a structured log.

```ruby
module StructuredTagComputer
  def self.call(request, taggers)
    {
      request_id: request.request_id
    }
  end
end

config.log_tag_computer = StructuredTagComputer
```

### Other Information

I'm interested in using the idea of Tagged Logging with Ougai, a structured/JSON logger. However, the default `TaggedLogging::Formatter` and `Rails::Rack::Logger.compute_tags` methods make assumptions that the logger will exclusively be logging strings.

This PR tackles the `compute_tags` part, making it default to the original implementation but allowing for it to be replaced with a custom callable.

In the end, I imagine the API would look like this:

```ruby
module StructuredTagFormatter
  # I'd like to extract the tagging parts of TaggedLogging::Formatter (everything except call and tags_text) 
  # in a future PR so that it can be included in custom Tagged Formatters
  include ActiveSupport::TaggedLogging::Taggable

  def call(severity, timestamp, progname, data)
    super(severity, timestamp, progname, merged_tags.merge(data))
  end
   
  private

  def merged_tags
    current_tags.reduce(:merge) || {}
  end
end

# config/environments/production.rb
config.log_tag_computer = StructuredTagComputer # see Summary

logger = Ougai::Logger.new(STDOUT)
# Passing in a custom tag formatter will be the other component of a future PR
config.logger = ActiveSupport::TaggedLogging.new(logger, StructuredTagFormatter)
```

SkipKayhil commented:
Hey @zzak, would you mind taking a look at this if you get a chance?

zzak commented:
@SkipKayhil Hey sorry I haven't had any cycles for this yet but I know RF was working on some structured logging stuff so it might be good to get his feedback! 🙏

skipkayhil committed:
make Rails::Rack::Logger tag compute configurable

For applications that use a structured/JSON logger, this is the first
change necessary to allow them to use the Tagged Logging feature in
Rails::Rack::Logger. Instead of returning an array of strings, a
configured tag computer could return a hash of fields to be merged into
a structured log.

SkipKayhil commented:
Hey @rafaelfranca, I would love some feedback on this. Whether you think this approach is reasonable or if you have suggestions for a better way forward, I'm very interested in better support for JSON logging in rails.

SkipKayhil commented:
🤦 Somehow it took me this long to realize this usage is already supported: https://github.com/rails/rails/blob/da91d1cd7575ff83791ef3494dcdf3033ffd2b97/railties/lib/rails/rack/logger.rb#L61-L62

```ruby
config.log_tags = [
  -> request do
    {
      trace_id: request.uuid
    }
  end
]
```

Sorry for the noise
